DevelopingSpace/starchart
# ci(webhooks): dockerized webhooks with tokens

# Description

This is a direct PR for #50 and provides a dockerized webhook service, along with simple token-based authentication.

A custom-made docker service has been added to the `docker-compose` file, called `webhook` and that contains the functionality for the webhook package. This is a Docker-in-Docker service therefore the host-side docker socket is mounted to this container so it can act upon running containers.

## Authentication
An `authenticate.sh` utility script is provided to handle authentication. By default, it will compare the values of the `DEFAULT_AUTH_TOKEN` (set through `docker-compose.yml`) and `HOOK_token` (set through `hooks.yml`) env vars, if the values match, it returns 0, otherwise a non-zero number. This can be used by webhook scripts to achieve authentication, like so
```sh
if .authenticate.sh; then   # the "token" query param must be the same as $DEFAULT_AUTH_TOKEN
        echo "You are in!"
else
        echo "Auth failed"
fi;
```
The authentication token can be overriden for webhooks so it uses a custom token and not the `DEFAULT_AUTH_TOKEN`. This can be done by providing the expected token as a parameter to the `authenticate.sh` utility script.
```sh
# to override: .authenticate.sh <token> <expected_token>
if .authenticate.sh $HOOK_token uniqueT0ken; then   # the "token" query param must be the same as uniqueT0ken
        echo "You are in!"
else
        echo "Auth failed"
fi;
```

# How To Test
1. Clone the repo
2. Run `docker compose up` (this will start the `webhook` service, and others)
3. Invoke the following command ```curl -X POST localhost:8000/hooks/deploy -H "Content-Type: application/json" -d '{"tag":"asdf", "service":"asfd", "image":"asdf", "token":"IAmASecretToken"}'``` (notice the token param is the same as the `DEFAULT_AUTH_TOKEN` env var in the `docker-compose.yml`)
4. Invoke the command ```curl -X POST localhost:8000/hooks/heartbeat -H "Content-Type: application/json" -d '{"token":"AnOverridenToken"}'```, notice how the token is different this time, this is because the heartbeat.sh script overrode the token like so 
```sh
# .authenticate.sh <token> <expected_token>
if ./.authenticate.sh $HOOK_TOKEN AnOverridenToken; then
  main
fi
```


https://user-images.githubusercontent.com/15690606/215421278-97c69c58-a351-45ee-b0bf-df549cf5825f.mp4
